presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -21,88 +22,376 @@
+                        SensitivityWatchEventModifier.HIGH);
+                isWatchServiceStarted.set(true);
+                log.info("Successfully registered watch service for %s", parentDir);
 
-        this.initializeServerWeights();
+                while (true) {
+                    WatchKey key = watchService.take();
+                    log.info("Changes to router config directory detected: %s", routerConfigFile);
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        log.info("Event detected: %s, path: %s", event.kind().name(), event.context());
+                        Path changed = (Path) event.context();
+                        if (changed.endsWith(routerConfigFile.getName())) {
+                            this.onConfigChangeDetection.apply();
+                        }
+                        else {
+                            log.info("Config change to %s ignored by ClusterManager (config file is %s)", event.context(), routerConfigFile.getName());
+                        }
+                    }
+                    key.reset();
+                }
+            }
+            catch (Exception e) {
+                log.error("Exception in file watcher loop while monitoring %s, %s", routerConfig.getConfigFile(), e);
+                throw new RuntimeException(e);
+            }
+        });
     }
 
     public List<URI> getAllClusters()
     {
-        return groups.values().stream()
+        return currentConfig.get().getGroups().values().stream()
                 .flatMap(groupSpec -> groupSpec.getMembers().stream())
                 .collect(toImmutableList());
     }
 
     public Optional<URI> getDestination(RequestInfo requestInfo)
     {
+        ClusterManagerConfig config = currentConfig.get();
         Optional<String> target = matchGroup(requestInfo);
         if (!target.isPresent()) {
             return Optional.empty();
         }
 
-        checkArgument(groups.containsKey(target.get()));
-        GroupSpec groupSpec = groups.get(target.get());
-        scheduler.setCandidates(groupSpec.getMembers());
-        if (schedulerType == WEIGHTED_RANDOM_CHOICE) {
-            scheduler.setWeights(serverWeights.get(groupSpec.getName()));
+        checkArgument(config.getGroups().containsKey(target.get()));
+        GroupSpec groupSpec = config.getGroups().get(target.get());
+
+        List<URI> healthyClusterURIs = groupSpec.getMembers().stream().filter((entry) -> {
+            if (remoteClusterInfos.containsKey(entry)) {
+                return remoteClusterInfos.get(entry).isHealthy();
+            }
+            else {
+                log.warn("Cluster [%s] is not available in remoteClusterInfos", entry);
+                return false;
+            }
+        }).collect(Collectors.toList());
+
+        if (healthyClusterURIs.isEmpty()) {
+            log.info("No healthy clusters found, enabled routing to unhealthy clusters");
```

Comment:
nit: `No healthy cluster found, will attempt to route using existing group spec`